### PR TITLE
Branch support

### PR DESCRIPTION
**Pull request by maximAtanasov**
Kanban borad: https://github.com/reflectoring/coderadar/projects/8

Fixes issues #1, #322

This PR adds support for additional branches in coderadar. The additional functionality works for the most part. Most of the Cypher queries have been rewritten and the REST-endpoints have been adjusted. The downside to all this is that things are a bit slower now (analysis, metric queries, saving).
Things can likely be optimized further.

Some of the notable changes in this PR are as follows:
 - A lot of Cypher-queries use APOC (https://neo4j.com/developer/neo4j-apoc/), which is needed for procedures cypher cannot do on it's own (most notably walking the commit tree up the `IS_CHILD_OF` relationships). This means everyone on the team will need to use Neo4j Desktop and install the APOC plugin, it takes only a couple of seconds to do so).
 - Files (and FileEntities) have two new attributes:
      - `sequenceId`: I added a asequence ID as a way to uniquely identify file nodes (the `ExtractProjectCommitsAdapter` will always generate the same sequenceId for the same File).
      - `objectHash` : This is the git object hash for the file. I believe this is the simplest way to identify renames/moves for the delta tree.
 - The `Start-/StopAnalyzingService` and `GetAnalzyingStatusService` have been consodilated into a single service called `AnalyzingService` which in turn implements three use-cases. This saves us two ports, two adapters, two database queries and we no longer have to perform an additional database query upon analyzing each commit.
 - The scan task scans for new branches and adds them to the project project.
 - Local project repositories aren't checked out anymore (as this isn't needed). This stops IntelliJ from going crazy and indexing everything.

There are also a few issues:

 - Saving a project with a non-null end date doesn't work as the HEAD commit of the project isn't saved. We should discuss whether or not users should ever be able to do this or if we should write extra code around this.
 - The city-view becomes buggy/broken when there are renames (maybe fixed in #323 ?).
 - Selecting a branch whose name begins with `#` will cause the UI to break.

However, I believe these issues are better left for a future PR as this one is large enough as it is. We could also look into integrating #269 once this is merged.

### CODE REVIEW

File: coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/projectadministration/module/repository/ModuleRepository.java [hunk 1]
```diff
@@ -9,41 +9,69 @@
...
   @Query(
-      "CREATE (m:ModuleEntity { path: {1} }) WITH m "
-          + "MATCH (p)-[r1:CONTAINS]->(f:FileEntity) WHERE ID(p) = {0} AND f.path STARTS WITH {1} WITH DISTINCT m, p, f, r1 "
-          + "CREATE (m)-[r2:CONTAINS]->(f) WITH DISTINCT m,f,p,r1 "
-          + "DELETE r1 WITH m, p LIMIT 1 "
-          + "CREATE (p)-[r3:CONTAINS]->(m) RETURN m")
+      "MATCH (p)-[:CONTAINS*1..]->(m:ModuleEntity) USING SCAN m:ModuleEntity WHERE ID(p) = {0} RETURN m")
   @NonNull
-  ModuleEntity createModuleInProject(@NonNull Long projectId, @NonNull String modulePath);
+  List<ModuleEntity> findModulesInProject(@NonNull Long projectId);
```
[the remark this answers is not in the document]
Comment (maximAtanasov) [reply coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/projectadministration/module/repository/ModuleRepository.java]: Done

File: coderadar-vcs/src/main/java/io/reflectoring/coderadar/vcs/adapter/ExtractProjectCommitsAdapter.java [hunk 2]
```diff
@@ -90,9 +112,11 @@ private Commit mapRevCommitToCommit(RevCommit rc) {
    * @param git The git API object.
    * @param firstCommit The firstCommit of the repository.
    * @param files A HashMap containing files already created for the project.
+   * @param seqeunceId One element array with the current file sequence number.
    * @throws IOException Thrown if the commit tree cannot be walked.
    */
-  private void setFirstCommitFiles(Git git, Commit firstCommit, HashMap<String, List<File>> files)
+  private void setFirstCommitFiles(
+      Git git, Commit firstCommit, HashMap<String, List<File>> files, long[] seqeunceId)
```
[the remark this answers is not in the document]
Comment (maximAtanasov) [reply coderadar-vcs/src/main/java/io/reflectoring/coderadar/vcs/adapter/ExtractProjectCommitsAdapter.java]: Fixed

File: coderadar-vcs/src/main/java/io/reflectoring/coderadar/vcs/adapter/ExtractProjectCommitsAdapter.java [hunk 3]
```diff
@@ -90,9 +112,11 @@ private Commit mapRevCommitToCommit(RevCommit rc) {
    * @param git The git API object.
    * @param firstCommit The firstCommit of the repository.
    * @param files A HashMap containing files already created for the project.
+   * @param seqeunceId One element array with the current file sequence number.
```
[the remark this answers is not in the document]
Comment (maximAtanasov) [reply coderadar-vcs/src/main/java/io/reflectoring/coderadar/vcs/adapter/ExtractProjectCommitsAdapter.java]: Fixed

File: coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/projectadministration/domain/ProjectEntity.java [hunk 4]
```diff
@@ -60,4 +55,9 @@
   @ToString.Exclude
   @EqualsAndHashCode.Exclude
   private List<AnalyzerConfigurationEntity> analyzerConfigurations = new ArrayList<>();
+
+  @Relationship(type = "HAS_BRANCH")
+  @ToString.Exclude
+  @EqualsAndHashCode.Exclude
```
[the remark this answers is not in the document]
Comment (maximAtanasov) [reply coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/projectadministration/domain/ProjectEntity.java]: Deleted

File: coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/analyzer/repository/MetricRepository.java [hunk 5]
```diff
@@ -11,14 +12,49 @@
...
+          + "CALL apoc.path.subgraphNodes(c, {relationshipFilter:'IS_CHILD_OF>'}) YIELD node WITH node as c ORDER BY c.timestamp DESC WITH collect(c) as commits "
+          + "CALL apoc.cypher.run('UNWIND commits as c OPTIONAL MATCH (f)<-[:RENAMED_FROM]-()-[:CHANGED_IN]->(c) RETURN collect(f) as renames', {commits: commits}) "
+          + "YIELD value WITH commits, value.renames as renames "
+          + "CALL apoc.cypher.run('UNWIND commits as c OPTIONAL MATCH (f)-[:CHANGED_IN {changeType: \"DELETE\"}]->(c) "
+          + "RETURN collect(f) as deletes', {commits: commits}) YIELD value WITH commits, renames, value.deletes as deletes "
+          + "UNWIND commits as c "
+          + "MATCH (f)-[:MEASURED_BY]->(m)-[:VALID_FOR]->(c) WHERE "
+          + "NOT(f IN deletes OR f IN renames) AND m.value <> 0 WITH ID(f) as id, m.name as name, head(collect(m)) as metric "
+          + "RETURN  id, collect(metric) as metrics")
+  List<LinkedHashMap<Object, Object>> getLastMetricsForFiles(Long projectId, String branchName);
```
[the remark this answers is not in the document]
Comment (maximAtanasov) [reply coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/analyzer/repository/MetricRepository.java]: Done

File: coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/projectadministration/domain/BranchEntity.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+@Data
+@NoArgsConstructor
+@NodeEntity
+public class BranchEntity {
+  private Long id;
+  private String name;
+  private String commitHash;
+
+  @Relationship(type = "POINTS_TO")
+  private CommitEntity commit;
```
[the remark this answers is not in the document]
Comment (maximAtanasov) [reply coderadar-graph/src/main/java/io/reflectoring/coderadar/graph/projectadministration/domain/BranchEntity.java]: Deleted

File: coderadar-core/src/main/java/io/reflectoring/coderadar/projectadministration/domain/Branch.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package io.reflectoring.coderadar.projectadministration.domain;
+
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+/** This class represents a branch in a project. */
+@Data
+@NoArgsConstructor
+public class Branch {
+  private long id;
```
[the remark this answers is not in the document]
Comment (maximAtanasov) [reply coderadar-core/src/main/java/io/reflectoring/coderadar/projectadministration/domain/Branch.java]: Deleted

File: coderadar-core/src/main/java/io/reflectoring/coderadar/vcs/port/driver/update/UpdateRepositoryUseCase.java [hunk 8]
```diff
@@ -1,6 +1,8 @@
 package io.reflectoring.coderadar.vcs.port.driver.update;
 
+import io.reflectoring.coderadar.projectadministration.domain.Branch;
 import io.reflectoring.coderadar.vcs.UnableToUpdateRepositoryException;
+import java.util.List;
 
 public interface UpdateRepositoryUseCase {
```
[the remark this answers is not in the document]
Comment (maximAtanasov) [reply coderadar-core/src/main/java/io/reflectoring/coderadar/vcs/port/driver/update/UpdateRepositoryUseCase.java]: Done

File: coderadar-core/src/main/java/io/reflectoring/coderadar/vcs/port/driven/UpdateRepositoryPort.java [hunk 9]
```diff
@@ -1,7 +1,9 @@
 package io.reflectoring.coderadar.vcs.port.driven;
 
+import io.reflectoring.coderadar.projectadministration.domain.Branch;
 import io.reflectoring.coderadar.vcs.UnableToUpdateRepositoryException;
 import io.reflectoring.coderadar.vcs.port.driver.update.UpdateRepositoryCommand;
+import java.util.List;
 
 public interface UpdateRepositoryPort {
```
[the remark this answers is not in the document]
Comment (maximAtanasov) [reply coderadar-core/src/main/java/io/reflectoring/coderadar/vcs/port/driven/UpdateRepositoryPort.java]: Done

File: coderadar-core/src/main/java/io/reflectoring/coderadar/projectadministration/port/driven/branch/DeleteBranchPort.java [hunk 10]
```diff
@@ -0,0 +1,3 @@
+package io.reflectoring.coderadar.projectadministration.port.driven.branch;
+
+public interface DeleteBranchPort {}
```
[the remark this answers is not in the document]
Comment (maximAtanasov) [reply coderadar-core/src/main/java/io/reflectoring/coderadar/projectadministration/port/driven/branch/DeleteBranchPort.java]: Done

File: coderadar-core/src/main/java/io/reflectoring/coderadar/projectadministration/port/driven/branch/CheckoutBranchPort.java [hunk 11]
```diff
@@ -0,0 +1,3 @@
+package io.reflectoring.coderadar.projectadministration.port.driven.branch;
+
+public interface CheckoutBranchPort {}
```
[the remark this answers is not in the document]
Comment (maximAtanasov) [reply coderadar-core/src/main/java/io/reflectoring/coderadar/projectadministration/port/driven/branch/CheckoutBranchPort.java]: Done

**Comment by maximAtanasov** [comment]
@KilianKrause Metrics should be correct now, but it would be nice if you could still check them.

File: coderadar-rest/src/main/java/io/reflectoring/coderadar/rest/query/GetCommitsInProjectController.java [hunk 12]
```diff
@@ -22,20 +20,12 @@ public GetCommitsInProjectController(GetCommitsInProjectUseCase getCommitsInProj
...
-              .setAnalyzed(commit.isAnalyzed())
-              .setAuthor(commit.getAuthor())
-              .setComment(commit.getComment())
-              .setTimestamp(commit.getTimestamp()));
-    }
-    return new ResponseEntity<>(responses, HttpStatus.OK);
+  @GetMapping(
+      path = "/projects/{projectId}/{branchName}/commits",
+      produces = MediaType.APPLICATION_JSON_VALUE)
+  public ResponseEntity<List<Commit>> listCommits(
```
Comment (KilianKrause) [inline coderadar-rest/src/main/java/io/reflectoring/coderadar/rest/query/GetCommitsInProjectController.java]: `GetCommitResponse` is no longer used anymore. We can remove this class and replace its occurrences in the tests with the `Commit` class.

**Comment by KilianKrause** [comment]
> @KilianKrause Metrics should be correct now, but it would be nice if you could still check them.

I've tested it with a few different projects, the metrics are correct.

**Pull request merged**